src/server/pps/server/master.go

```diff
@@ -158,6 +166,31 @@ eventLoop:
 	}
 }
 
+func (m *ppsMaster) attemptStep(ctx context.Context, e *pipelineEvent) error {
+	var errCount int
+	return backoff.RetryNotify(func() error {
+		// Create/Modify/Delete pipeline resources as needed per new state
+		return m.step(e.pipeline, e.etcdVer, e.etcdRev)
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+		var failError failPipelineError
+		if errors.As(err, &failError) {
+			// don't just return this error, we might have to fail the pipeline
+			if errCount++; errCount >= maxErrCount || !failError.retry {
+				failError := m.a.setPipelineFailure(ctx, e.pipeline, fmt.Sprintf(
+					"could not update resources after %d attempts: %v", errCount, err))
+				if failError != nil {
+					return fmt.Errorf("error failing pipeline %q: %v", e.pipeline, failError)
```

Comment:
In general I think we're preferring `error.Wrapf` over `fmt.Errorf(..., err)` (if that works for this code? I think we've been preferring it because it works better with `errors.As`, but maybe that's a problem here?)

Reply:
Sounds good, I don't think it should affect functionality